programs/marinade-finance/src/instructions/management/redelegate.rs

```diff
@@ -0,0 +1,368 @@
+        // compute total target stake (current total active stake +/- delta)
+        let total_stake_target_i128 =
+            self.state.validator_system.total_active_balance as i128 + total_stake_delta_i128;
+        // convert to u64
+        let total_stake_target =
+            u64::try_from(total_stake_target_i128).expect("total_stake_target+stake_delta");
+        // compute target for this particular validator (total_stake_target * score/total_score)
+        let validator_stake_target = self
+            .state
+            .validator_system
+            .validator_stake_target(&validator, total_stake_target)?;
+        // if validator is already on-target (or the split will be lower than min_stake), exit now
+        if validator.active_balance <= validator_stake_target + self.state.stake_system.min_stake {
+            msg!(
+                "Current validator {} stake {} is <= target {} +min_stake",
+                validator.validator_account,
+                validator.active_balance,
+                validator_stake_target
+            );
+            return Ok(()); // Not an error. Don't fail other instructions in tx
+        }
+
+        // compute how much we can unstake from this validator, and cap unstake amount to it
+        let max_redelegate_from_validator = validator.active_balance - validator_stake_target;
+        let re_delegate_amount = if desired_redelegate_amount > max_redelegate_from_validator {
+            max_redelegate_from_validator
+        } else {
+            desired_redelegate_amount
+        };
+        // compute how much this particular account will have after split
+        let stake_account_after = stake
+            .last_update_delegated_lamports
+            .saturating_sub(re_delegate_amount);
+
+        // get dest validator from index
+        let mut dest_validator = self.state.validator_system.get(
+            &self.validator_list.data.as_ref().borrow(),
+            dest_validator_index,
+        )?;
+        // verify dest_validator_account matches
+        if self.dest_validator_account.key() != dest_validator.validator_account {
+            return err!(MarinadeError::IncorrectDestValidatorAccountOrIndex);
+        }
+
+        // compute dest validator target
+        let dest_validator_stake_target = self
+            .state
+            .validator_system
+            .validator_stake_target(&dest_validator, total_stake_target)?;
+        // if added stake will put the validator over target, reject
+        if dest_validator.active_balance + re_delegate_amount > dest_validator_stake_target {
+            msg!(
+                "Dest validator {} stake {} + amount {} is > target {}",
+                dest_validator.validator_account,
+                dest_validator.active_balance,
+                re_delegate_amount,
+                dest_validator_stake_target
+            );
+            return err!(MarinadeError::RedelegateOverTarget);
+        }
+
```

Comment:
After final calculation of re_delegate_amount it must be checked to be >= minStake (when you changes the calculation). Otherwise return Ok(())